compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -179,6 +200,19 @@ impl<'a> Parser<'a> {
         &self.sess.span_diagnostic
     }
 
+    pub(super) fn restore(&mut self, snapshot: SnapshotParser<'a>) {
+        *self = snapshot.parser;
+        self.unclosed_delims.extend(snapshot.unclosed_delims.clone());
+    }
+
+    pub(super) fn diagnostic_snapshot(&self) -> SnapshotParser<'a> {
+        let mut snapshot = self.clone();
+        let unclosed_delims = self.unclosed_delims.clone();
+        // initialize unclosed_delims to avoid duplicate errors.
```

Comment:
```suggestion
        // Clear `unclosed_delims` in snapshot to avoid
        // duplicate errors being emitted when the `Parser`
        // is dropped (which may or may not happen, depending
        // if the parsing the snapshot is created for is successful)
```